topostats/grainstats.py

```diff
@@ -0,0 +1,653 @@
+        """Class method that takes a 2D boolean numpy array image of a grain and returns a python list of the
+        coordinates of the edges of the grain.
+
+        Parameters
+        ----------
+        grain_mask : np.ndarray
+            A 2D numpy array image of a grain. Data in the array must be boolean.
+
+        Returns
+        -------
+        edges : list
+            List containing the coordinates of the edges of the grain.
+        """
+        # Fill any holes
+        filled_grain_mask = scipy.ndimage.binary_fill_holes(grain_mask)
+        # Get outer edge using canny filtering
+        edges = skimage_feature.canny(filled_grain_mask, sigma=3)
+        nonzero_coordinates = edges.nonzero()
+        # Get vector representation of the points
+        # FIXME : Switched to list comprehension but should be unnecessary to create this as a list as we can use
+        # np.stack() to combine the arrays and use that...
+        # return np.stack(nonzero_coordinates, axis=1)
+        # edges = []
+        # for vector in np.transpose(nonzero_coordinates):
+        #     edges.append(list(vector))
+        # return edges
+        return [list(vector) for vector in np.transpose(nonzero_coordinates)]
+
+    def calculate_radius_stats(self, edges: list) -> Tuple:
+        """Class method that calculates the statistics relating to the radius. The radius in this context
+        is the distance from the centroid to points on the edge of the grain.
+
+        Parameters
+        ----------
+        edges: list
+            A 2D python list containing the coordinates of the edges of a grain.
+
+        Returns
+        -------
+        Tuple[float]
+            A tuple of the minimum, maximum, mean and median radius of the grain
+        """
+        # Convert the edges to the form of a numpy array
+        # edges = np.array(edges)
+        # Calculate the centroid of the grain
+        centroid = self._calculate_centroid(edges)
+        # Calculate the displacement
+        displacements = self._calculate_displacement(edges, centroid)
+        # Calculate the radius of each point
+        radii = self._calculate_radius(displacements)
+        return {"min": np.min(radii), "max": np.max(radii), "mean": np.mean(radii), "median": np.median(radii)}
+
+    def _calculate_centroid(self, edges: np.array) -> tuple:
+        """Calculate the centroid of a bounding box.
+
+        Parameters
+        ----------
+        edges: list
+            A 2D python list containing the co-ordinates of the edges of a grain.
+
+        Returns
+        -------
+        tuple
+            The co-ordinates of the centroid.
+        """
+        # FIXME : Remove once we have a numpy array returned by calculate_edges
+        edges = np.array(edges)
+        return (sum(edges[:, 0]) / len(edges), sum(edges[:, 1] / len(edges)))
+
+    def _calculate_displacement(self, edges: np.array, centroid: tuple) -> np.array:
```

Comment:
Could be confused with `get_displacement`?

Reply:
Renamed `get_displacement` > `calculate_squared_distance`.